### ISSUE TITLE
UNet with Pascal VOC gives 5% mIOU

### ISSUE THREAD

**Issue by RishiTejaMadduri** [yassouali/pytorch_segmentation#86]
Hi, I have been trying to run UNet with Pascal VOC dataset and I am getting really bad results. The same dataset is doing great with PSPNet but UNet results are really bad. 

Here is my config: 

```
{
    "name": "UNET",
    "n_gpu": 1,
    "use_synch_bn": true,

    "arch": {
        "type": "UNet",
        "args": {
            "backbone": "resnet50",
            "freeze_bn": false,
            "freeze_backbone": false
        }
    },

    "train_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/mnt/batch/tasks/shared/LS_root/mounts/clusters/objloc/code/datasets/VOC/",
            "batch_size": 8,
            "base_size": 400,
            "crop_size": 380,
            "augment": true,
            "shuffle": true,
            "scale": true,
            "flip": true,
            "rotate": true,
            "blur": false,
            "split": "train",
            "num_workers": 8
        }
    },

    "val_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/mnt/batch/tasks/shared/LS_root/mounts/clusters/objloc/code/datasets/VOC/",
            "batch_size": 8,
            "crop_size": 480,
            "val": true,
            "split": "val",
            "num_workers": 4
        }
    },

    "optimizer": {
        "type": "SGD",
        "differential_lr": true,
        "args":{
            "lr": 0.001,
            "weight_decay": 0.0001,
            "momentum": 0.9
        }
    },

    "loss": "CrossEntropyLoss2d",
    "ignore_index": 255,
    "lr_scheduler": {
        "type": "Poly",
        "args": {}
    },

    "trainer": {
        "epochs": 80,
        "save_dir": "/mnt/batch/tasks/shared/LS_root/mounts/clusters/objloc/code/pytorch_segmentation/saved/",
        "save_period": 10,
  
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        
        "tensorboard": true,
        "log_dir": "saved/runs",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}
```

am I doing something wrong?

**Comment by yassouali** [comment]
Hi @RishiTejaMadduri 

Since for Unet, you are training from scratch (no trained backbone), I would use a larger learning rate (0.1 for example), and see if it helps. You can also avoid using sync batch norm, since it slows the training and only necessary for multi GPUs.

**Comment by pedroramirez3** [comment]
hi @yassouali
first of all, amazing project!
i'm having the same prob as the OP.. do you happen to have a pretrained UNet you can share? 
it'd be great to run some evaluations, i'd really appreciate it!

or, alternatively, do you have an example config file for a UNet that leads to a successful training?

**Comment by yassouali** [comment]
HI @pedroramirez3 (and @RishiTejaMadduri )

I just added a unet version with a resnet backbone and unet style decoder. This version will be a lot easier to train since the encoder is more powerful and is already pretrained. 

You can simply replace PSPNet in the config.json file by UNetResnet (maybe also test training the model with ADAM, it might help a bit more). Good luck.

**Comment by pedroramirez3** [comment]
hi @yassouali 
thanks for your advice!
however, i really need to keep the original unet architecture
any pointers in that direction?

btw, i found a bug in the BaseTrainer class, where **_resume_checkpoint** is expecting a logger in the checkpoint but **_save_checkpoint** is not saving one (opening a new issue)

**Comment by yassouali** [comment]
@pedroramirez3 sorry for the delay.

I just did go through the unet architecture and fixed some parts. I think the current version is in line with other implementations, hopefully, this version works for you.

**Comment by yassouali** [comment]
closed for now, please feel free to open it if you have other questions

**Comment by gauridhawan** [comment]
Hi, we tried using a similar config with learning rate 0.1 and Adam optimizer but it still gives mIoU as 0.3, is there a way to improve it?

**Comment by FersSky** [comment]
Hello, I'm also interested in this project because it is very comprehensive. My configuration.json file is quite similar I'm using 2 GPUs the data is also VOC, and even though I have made the adjustments mentioned here, the MIoU does not improve beyond 0.04, even though the loss function shows improvement. Here is one of my results:

TRAIN (3) | Loss: 1.210 | Acc 0.71 | mIoU 0.04 | B 0.72 | D 0.02 |

In my case, I'm using the SGD optimizer with a learning rate of 0.1 and a momentum of 0.9, and my loss function is "CrossEntropyLoss2d". I also tried training with Adam, but these settings did not improve the results either.

Has anyone managed to achieve better performance?
